### PR TITLE
Implementation of `matmul`, `tensordot`, and `vecdot` per array API

### PR DESCRIPTION
**Pull request by ndgrigorian**
This pull request implements `dpctl.tensor.matmul`, `dpctl.tensor.tensordot`, and `dpctl.tensor.vecdot` functions as per array API specification, as well as GEMM and dot product kernels necessary for these implementations.

A new `dpctl.tensor` submodule `_tensor_linalg_impl` has been added, where the common Python binding `py_dot` is implemented, and may be used to house future linear algebra implementations.

This pull request also includes a change to remove an unnecessary file, `elementwise_common.cpp`, left behind when elementwise functions were moved to a separate submodule.

- [X] Have you provided a meaningful PR description?
- [X] Have you added a test, reproducer or referred to an issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [X] Have you checked performance impact of proposed changes?
- [X] If this PR is a work in progress, are you opening the PR as a draft?

Prerequisite for merging:

- [x] Implement Python operators `__matmul__` and `__imatmul__` in `usm_ndarray`.
- [x] Review gemm tree-reduction kernels, implement corner cases to address assertions for certain input sizes
- [x] Improve coverage

### CODE REVIEW

File: dpctl/tensor/_linear_algebra_functions.py [hunk 1]
```diff
@@ -46,3 +62,921 @@ def matrix_transpose(x):
...
+
+            The data type of the returned array is determined by the Type
+            Promotion Rules. If either `x1` or `x2` has a complex floating
+            point type, neither argument is complex conjugated or transposed.
+    """
+    if not isinstance(x1, dpt.usm_ndarray):
+        raise TypeError(f"Expected dpctl.tensor.usm_ndarray, got {type(x1)}")
+    if not isinstance(x2, dpt.usm_ndarray):
+        raise TypeError(f"Expected dpctl.tensor.usm_ndarray, got {type(x2)}")
+    if order not in ["K", "C", "F", "A"]:
```
Comment (vtavana) [inline dpctl/tensor/_linear_algebra_functions.py:643]: Do you want to change `order` even if a user uses the lower case letters?

File: dpctl/tensor/_linear_algebra_functions.py [hunk 2]
```diff
@@ -46,3 +62,921 @@ def matrix_transpose(x):
...
+                raise ValueError(
+                    f"`matmul` input `x1` cannot be cast from {x1_dtype} to "
+                    f"requested type {res_dt} according to the casting rule "
+                    "''same_kind''."
+                )
+        if x2_dtype != res_dt:
+            if dpt.can_cast(x2_dtype, res_dt, casting="same_kind"):
+                buf2_dt = res_dt
+            else:
+                raise ValueError(
```
Comment (vtavana) [inline dpctl/tensor/_linear_algebra_functions.py:728]:
```suggestion
                raise TypeError(
```

File: dpctl/tensor/_linear_algebra_functions.py [hunk 3]
```diff
@@ -46,3 +62,921 @@ def matrix_transpose(x):
...
+            )
+    else:
+        res_dt = dpt.dtype(dtype)
+        res_dt = _to_device_supported_dtype(res_dt, sycl_dev)
+        buf1_dt, buf2_dt = None, None
+        if x1_dtype != res_dt:
+            if dpt.can_cast(x1_dtype, res_dt, casting="same_kind"):
+                buf1_dt = res_dt
+            else:
+                raise ValueError(
```
Comment (vtavana) [inline dpctl/tensor/_linear_algebra_functions.py:719]:
```suggestion
                raise TypeError(
```

File: dpctl/tensor/_linear_algebra_functions.py [hunk 4]
```diff
@@ -46,3 +62,921 @@ def matrix_transpose(x):
...
+
+        if out.shape != res_shape:
+            raise ValueError(
+                "The shape of input and output arrays are inconsistent. "
+                f"Expected output shape is {res_shape}, got {out.shape}"
+            )
+
+        if res_dt != out.dtype:
+            raise ValueError(
+                f"Output array of type {res_dt} is needed," f"got {out.dtype}"
```
Comment (vtavana) [inline dpctl/tensor/_linear_algebra_functions.py:749]:
```suggestion
            raise TypeError(
                f"Output array of type {res_dt} is needed, " f"got {out.dtype}"
```

**Review by vtavana — COMMENTED** [review]
I ran the tests that we have in `dpnp` for `matmul` using `dpt.matmul` and they were passed.

File: dpctl/tensor/_linear_algebra_functions.py (hunk 2 above)
Comment (oleksandr-pavlyk) [reply dpctl/tensor/_linear_algebra_functions.py:728]: Definitely not a `TypeError`. See https://docs.python.org/3/library/exceptions.html#TypeError, specifically note:

> Passing arguments of the wrong type (e.g. passing a [list](https://docs.python.org/3/library/stdtypes.html#list) when an [int](https://docs.python.org/3/library/functions.html#int) is expected) should result in a [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError), but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError).

The input here is `usm_ndarray`, so the the type of input is as expected, but the value of input (it's internal property, `dtype`) is not as expected, hence `ValueError` is warranted.

File: dpctl/tensor/_linear_algebra_functions.py (hunk 1 above)
Comment (oleksandr-pavlyk) [reply dpctl/tensor/_linear_algebra_functions.py:643]: I do not think it is worth the bother to support lower case here. Documentation states upper case is what is expected.

**Comment by oleksandr-pavlyk** [comment]
> I ran the tests that we have in `dpnp` for `matmul` using `dpt.matmul` and they passed.

Thank you for running these checks @vtavana !

**Review by oleksandr-pavlyk — APPROVED** [review]
LGTM @ndgrigorian ! Thanks for making it happen. All additional work can be done in new PRs.

**Pull request merged**